libp2p-xtra/src/node.rs

```diff
@@ -0,0 +1,482 @@
+    ///
+    /// The given protocols will be tried **in order**, with the first successful one being used.
+    /// Specifying multiple protocols can useful to maintain backwards-compatibility. A node can
+    /// attempt to first establish a substream with a new protocol and falling back to older
+    /// versions in case the new version is not supported.
+    pub fn multiple_protocols(peer: PeerId, protocols: Vec<&'static str>) -> Self {
+        Self {
+            peer,
+            protocols,
+            marker_num_protocols: PhantomData,
+        }
+    }
+}
+
+/// Connect to the given [`Multiaddr`].
+///
+/// The address must contain a `/p2p` suffix.
+/// Will fail if we are already connected to the peer.
+#[derive(Debug)]
+pub struct Connect(pub Multiaddr);
+
+/// Disconnect from the given peer.
+#[derive(Clone, Copy, Debug)]
+pub struct Disconnect(pub PeerId);
+
+/// Listen on the provided [`Multiaddr`].
+///
+/// For this to work, the [`Node`] needs to be constructed with a compatible transport.
+/// In other words, you cannot listen on a `/memory` address if you haven't configured a `/memory`
+/// transport.
+pub struct ListenOn(pub Multiaddr);
+
+/// Retrieve [`ConnectionStats`] from the [`Node`].
+#[derive(Clone, Copy, Debug)]
+pub struct GetConnectionStats;
+
+#[derive(Debug)]
+pub struct ConnectionStats {
+    pub connected_peers: HashSet<PeerId>,
+    pub listen_addresses: HashSet<Multiaddr>,
+}
+
+/// Notifies an actor of a new, inbound substream from the given peer.
+#[derive(Debug)]
+pub struct NewInboundSubstream {
+    pub peer: PeerId,
+    pub stream: Substream,
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("No connection to {0}")]
+    NoConnection(PeerId),
+    #[error("Timeout in protocol negotiation")]
+    NegotiationTimeoutReached,
+    #[error("Failed to negotiate protocol")]
+    NegotiationFailed(#[from] NegotiationError), // TODO(public-api): Consider breaking this up.
+    #[error("Bad connection")]
+    BadConnection(#[from] yamux::ConnectionError), // TODO(public-api): Consider removing this.
+    #[error("Address {0} does not end with a peer ID")]
+    NoPeerIdInAddress(Multiaddr),
+    #[error("Either currently connecting or already connected to peer {0}")]
+    AlreadyConnected(PeerId),
+}
+
+impl Node {
+    /// Construct a new [`Node`] from the provided transport.
+    ///
+    /// A [`Node`]s identity ([`PeerId`]) will be computed from the given [`Keypair`].
+    ///
+    /// The `connection_timeout` is applied to:
+    /// 1. Connection upgrades (i.e. noise handshake, yamux upgrade, etc)
+    /// 2. Protocol negotiations
+    ///
+    /// The provided substream handlers are actors that will be given the fully-negotiated
+    /// substreams whenever a peer opens a new substream for the provided protocol.
+    pub fn new<T, const N: usize>(
+        transport: T,
+        identity: Keypair,
+        connection_timeout: Duration,
+        inbound_substream_handlers: [(
```

Comment:
Would it be worth making this argument a hashmap to let the user know each handler should be unique. It has to be converted to a hashmap anyway?

Reply:
I've chosen to make it an array because I found it more ergonomic to construct. But yes, I may need some documentation that points out that these have to be unique mappings.

Reply:
Should probably fail if we find a duplicate one.

Reply:
Added a check and a test for it.